diff --git a/libavformat/http.c b/libavformat/http.c
index 6af3259a43..f76a71fbb7 100644
--- a/libavformat/http.c
+++ b/libavformat/http.c
@@ -1298,7 +1298,7 @@ static int http_buf_read(URLContext *h, uint8_t *buf, int size)
             s->chunksize = strtoull(line, NULL, 16);
 
             av_log(h, AV_LOG_TRACE,
-                   "Chunked encoding data size: %"PRIu64"'\n",
+                   "Chunked encoding data size: %"PRIu64"\n",
                     s->chunksize);
 
             if (!s->chunksize) {
